Add data packing builtin functions

dneto0 opened the pull request:
* Add placeholder for unpacking builtin functions
* Expand the description of floating point conversions now that
  we have to handle edge cases for f32 -> f16.
  I put in the behaviour of an NVIDIA GPU for the concerning edge
  case, but added an Issue to review this for portability.

dneto0 commented:
Partly addresses #988

dneto0 commented:
HTML preview is at https://gistpreview.github.io/?55aec7912b75ea316e79597eef8a5053

dj2 commented on wgsl/index.bs:
```diff
@@ -4242,6 +4250,69 @@ TODO(dsinclair): Need gather operations
+        16 &times; |i| + 15 of the result.
+
+  <tr algorithm="packing 2xf16">
+    <td class="nowrap">`pack2xf16`(|e|: vec2&lt;f32&gt;) -> u32
```
This ones name seems to be the odd one out. Either this should be `pack2x16float` to match the above, or could we change the above to be `pack4xu8`, `pack2xu16`, `pack2xi8`, `pack2xi16` which matches the `pack2xf16`. In theory, if we get the smaller types, then those names should all match.

dneto0 replied:
These pack (and later unpack) functions are for doing in value types what happens when reading/writing texels from storage textures.  So I wanted to match the texel channels. 

So you're right it should be pack2x16float.  I'll fix it.

dneto0 replied:
Done

dneto0 commented:
Rebased and addressed @dj2's comment.

The main issue is the addition of general language about converting on floating point type value to another floating point type value.  The interesting case if f32 to f16 as it introduces cases where finite source numbers should map to either infinite f16 or max normal f16 values.  It's not clear what the dividing line is.  I've added an "Issue" to verify behaviour and revisit in the future.

kvark commented on wgsl/index.bs:
```diff
@@ -4242,6 +4250,69 @@ TODO(dsinclair): Need gather operations
+
+  <tr algorithm="packing 2x16unorm">
+    <td class="nowrap">`pack2x16unorm`(|e|: vec2&lt;f32&gt;) -> u32
+    <td>Converts two normalized floating point values to 8-bit unsigned integers, and then combines them
```
this should say "16-bit"

kvark reviewed (commented):
Do we have an investigation somewhere showing how these map to shading languages?

grorg commented:
Discussed at [2020-11-03 meeting](https://docs.google.com/document/d/1oV6pmwFOcomeuInQRl-erFt6ZiQfApXYSjoJis7zetA/).

grorg commented:
Resolved: Accept PR (although @kvark mentioned a small error)

dneto0 commented:
UGH.  I made a mistake claiming these were already present 1-1 in all the backends.  I had checked Metal but skimped on HLSL. (!)

> Do we have an investigation somewhere showing how these map to shading languages?

I'll add details here.

dneto0 commented:
- WGSL `pack2x16float(v :  vec2<f32>) -> u32`
    - SPIR-V GLSL.std.450 extended instruction PackHalf2x16
    - MSL:   `as_type<uint>(packed_half2(v.x, v.y))`  
    - HLSL:      `f32tof16(f.x) | (f32tof16(f.y) << 16)`

- WGSL: `pack4x8snorm(e: vec4<f32>)`
    - SPIR-V GLSL.std.450 extended instruction PackSnorm4x8
    - MSL:  `pack_float_to_snorm4x8(v)`
    - HLSL:  `uint4 temp = round(clamp(v, -1.0f, 1.0f)*127.0);  
                     return (temp.x & 0xff) | ((temp.y&0xff) <<8)   | ((temp.z&0xff) <<16)    | ((temp.w & 0xff) <<24);  `

- WGSL: `pack4x8unorm(e: vec4<f32>)`
    - SPIR-V GLSL.std.450 extended instruction PackUnorm4x8
    - MSL:  `pack_float_to_unorm4x8(v)`
    - HLSL:  `uint4 temp = round(clamp(v, 0.0f, 1.0f)*255.0);  
                     return (temp.x & 0xff) | ((temp.y&0xff) <<8)   | ((temp.z&0xff) <<16)    | ((temp.w & 0xff) <<24);  `

- WGSL: `pack2x16snorm(e: vec4<f32>)`
    - SPIR-V GLSL.std.450 extended instruction PackSnorm2x16
    - MSL:  `pack_float_to_snorm2x16(v)`
    - HLSL:  `uint4 temp = round(clamp(v, -1.0f, 1.0f)*32767.0);  
                     return (temp.x & 0xffff) | ((temp.y&0xffff) <<16);  `

- WGSL: `pack2x16unorm(e: vec4<f32>)`
    - SPIR-V GLSL.std.450 extended instruction PackUnorm2x16
    - MSL:  `pack_float_to_unorm2x16(v)`
    - HLSL:  `uint4 temp = round(clamp(v, 0.0f, 1.0f)*65535.0);  
                     return (temp.x & 0xffff) | ((temp.y&0xffff) <<16);  `

- WGSL: `pack2x16unorm(e: vec4<f32>)`
    - SPIR-V GLSL.std.450 extended instruction PackUnorm2x16
    - MSL:  `pack_float_to_unorm2x16(v)`
    - HLSL:  `uint4 temp = round(clamp(v, 0.0f, 1.0f)*65535.0);  
                     return (temp.x & 0xffff) | ((temp.y&0xffff) <<16);  `

dneto0 commented:
I filed #1201 to track the edge case behaviour of conversion from f32 to f16

grorg commented:
Discussed at the [2020-11-10 meeting](https://docs.google.com/document/d/1QFWUKpCcoLCMGHfJQWJurSPk-9PDcJ1Gc51FSg3_Fr8/).

Resolved: Accept with @dneto0 's changes.

dneto0 commented:
> Resolved: Accept with @dneto0 's changes.

The "changes" are really implementor's notes. The behaviour is already fully specified in the WGSL spect text in the PR.
